src/AspNetCore/Src/applicationinfo.cpp

```diff
@@ -0,0 +1,330 @@
+    APP_OFFLINE_HTM *pOldAppOfflineHtm = NULL;
+    APP_OFFLINE_HTM *pNewAppOfflineHtm = NULL;
+
+    if (INVALID_FILE_ATTRIBUTES == GetFileAttributes(strFilePath.QueryStr()) && GetLastError() == ERROR_FILE_NOT_FOUND)
+    {
+        m_fAppOfflineFound = FALSE;
+    }
+    else
+    {
+        m_fAppOfflineFound = TRUE;
+        pNewAppOfflineHtm = new APP_OFFLINE_HTM(strFilePath.QueryStr());
+
+        if (pNewAppOfflineHtm != NULL)
+        {
+            if (pNewAppOfflineHtm->Load())
+            {
+                //
+                // loaded the new app_offline.htm
+                //
+                pOldAppOfflineHtm = (APP_OFFLINE_HTM *)InterlockedExchangePointer((VOID**)&m_pAppOfflineHtm, pNewAppOfflineHtm);
+
+                if (pOldAppOfflineHtm != NULL)
+                {
+                    pOldAppOfflineHtm->DereferenceAppOfflineHtm();
+                    pOldAppOfflineHtm = NULL;
+                }
+            }
+            else
+            {
+                // ignored the new app_offline file because the file does not exist.
+                pNewAppOfflineHtm->DereferenceAppOfflineHtm();
+                pNewAppOfflineHtm = NULL;
+            }
+        }
+
+        // recycle the application
+        if (m_pApplication != NULL)
+        {
+            m_pApplication->ShutDown();
+            m_pApplication->DereferenceApplication();
+            m_pApplication = NULL;
+        }
+    }
+}
+
+HRESULT
+APPLICATION_INFO::EnsureApplicationCreated()
+{
+    HRESULT hr = S_OK;
+    BOOL    fLocked = FALSE;
+    APPLICATION* pApplication = NULL;
+    STACK_STRU(struFileName, 300);  // >MAX_PATH
+    STRU        hostFxrDllLocation;
+
+    if (m_pApplication != NULL)
+    {
+        goto Finished;
+    }
+
+    FindRequestHandlerAssembly();
+
+    if (m_pApplication == NULL)
+    {
+        AcquireSRWLockExclusive(&m_srwLock);
+        fLocked = TRUE;
+        if (m_pApplication != NULL)
+        {
+            goto Finished;
+        }
+
+        if (m_pfnAspNetCoreCreateApplication == NULL)
+        {
+            hr = HRESULT_FROM_WIN32(ERROR_INVALID_FUNCTION);
+            goto Finished;
+        }
+
+        hr = m_pfnAspNetCoreCreateApplication(m_pServer, m_pConfiguration, &pApplication);
+        if (FAILED(hr))
+        {
+            goto Finished;
+        }
+        m_pApplication = pApplication;
+    }
+
+Finished:
+    if (fLocked)
+    {
+        ReleaseSRWLockExclusive(&m_srwLock);
+    }
+    return hr;
+}
+
+HRESULT
+APPLICATION_INFO::FindRequestHandlerAssembly()
+{
+    HRESULT hr = S_OK;
+    BOOL    fLocked = FALSE;
+    STACK_STRU(struFileName, 256);
+
+    if (!g_fAspnetcoreRHAssemblyLoaded)
+    {
+        AcquireSRWLockExclusive(&g_srwLock);
+        fLocked = TRUE;
+        if (g_fAspnetcoreRHAssemblyLoaded)
+        {
+            goto Finished;
+        }
+
+        // load assembly and create the application
+        if (m_pConfiguration->QueryHostingModel() == APP_HOSTING_MODEL::HOSTING_IN_PROCESS)
```

Comment:
This logic will change based on the changes to hostfxr. The logic (for both in proc and out of proc) will be:
if hostfxrpath exists and hostfxr has the ability to find the requesthandler
   use hostfxr to find the request handler
otherwise if we are running out of process
   use the global requesthandler

Another issue is bitness for inprocess and out of process. If we are running a 32-bit worker process and find a 64 bit hostfxr, out of process can still work, but in-process cannot. However, if we want out-of-process to be loaded the same way we do in process, we would need a hostfxr.dll to be the same bitness as the worker process. We will need to discuss the finer details here.